File: starters/kotlin-android/core/src/main/kotlin/org/example/foundation/core/domain/Task.kt
// SPDX-FileCopyrightText: 2026 Zendrhax LLC
// SPDX-License-Identifier: MPL-2.0
package org.example.foundation.core.domain

const val MAX_TITLE_CODE_POINTS = 80
private val TASK_ID = Regex("[A-Za-z0-9_-]{1,100}")
private val TASK_TIME_RANGE = 0L..8_640_000_000_000_000L

enum class TaskError {
    INVALID_TITLE, TITLE_TOO_LONG, INVALID_ID, INVALID_TIME,
    NOT_FOUND, DUPLICATE_ID, STORAGE_FAILURE, DEPENDENCY_FAILURE,
}

sealed interface TaskResult<out T> {
    data class Success<T>(val value: T) : TaskResult<T>
    data class Failure(val error: TaskError) : TaskResult<Nothing>
}

data class Task(
    val id: String,
    val title: String,
    val completed: Boolean,
    val createdAtEpochMs: Long?,
) {
    init {
        require(TASK_ID.matches(id)) { "Invalid task identifier" }
        require(createdAtEpochMs == null || createdAtEpochMs in TASK_TIME_RANGE) { "Invalid creation time" }
        require(validateTitle(title) == TaskResult.Success(title)) { "Invalid or untrimmed task title" }
    }
}

private fun String.hasUnpairedSurrogate(): Boolean {
    var index = 0
    while (index < length) {
        val character = this[index]
        if (character.isHighSurrogate()) {
            if (index + 1 >= length || !this[index + 1].isLowSurrogate()) return true
            index += 2
        } else {
            if (character.isLowSurrogate()) return true
            index++
        }
    }
    return false
}

fun validateTitle(input: Any?): TaskResult<String> {
    if (input !is String) return TaskResult.Failure(TaskError.INVALID_TITLE)
    val title = input.trim()
    if (title.isEmpty() || title.hasUnpairedSurrogate() || title.any { it.isISOControl() || it == '\u2028' || it == '\u2029' }) {
        return TaskResult.Failure(TaskError.INVALID_TITLE)
    }
    if (title.codePointCount(0, title.length) > MAX_TITLE_CODE_POINTS) {
        return TaskResult.Failure(TaskError.TITLE_TOO_LONG)
    }
    return TaskResult.Success(title)
}

fun createTask(id: String, title: Any?, createdAtEpochMs: Long): TaskResult<Task> {
    val normalized = validateTitle(title)
    if (normalized is TaskResult.Failure) return normalized
    if (!TASK_ID.matches(id)) return TaskResult.Failure(TaskError.INVALID_ID)
    if (createdAtEpochMs !in TASK_TIME_RANGE) return TaskResult.Failure(TaskError.INVALID_TIME)
    return TaskResult.Success(Task(id, (normalized as TaskResult.Success).value, false, createdAtEpochMs))
}

fun toggleTask(task: Task): Task = task.copy(completed = !task.completed)

File: starters/web/src/domain/task.ts
export const MAX_TITLE_LENGTH = 80;

export type TaskErrorCode =
  | 'INVALID_TITLE'
  | 'TITLE_TOO_LONG'
  | 'INVALID_ID'
  | 'INVALID_TIMESTAMP'
  | 'NOT_FOUND'
  | 'DUPLICATE_ID'
  | 'STORAGE_UNAVAILABLE'
  | 'DEPENDENCY_FAILURE';

export type TaskError = Readonly<{ code: TaskErrorCode }>;
export type Result<T> =
  | Readonly<{ ok: true; value: T }>
  | Readonly<{ ok: false; error: TaskError }>;

export type Task = Readonly<{
  id: string;
  title: string;
  completed: boolean;
  createdAtEpochMs: number | null;
}>;

export function failure(code: TaskErrorCode): Result<never> {
  return { ok: false, error: { code } };
}

export function validateTitle(input: unknown): Result<string> {
  if (typeof input !== 'string') return failure('INVALID_TITLE');
  const title = input.trim();
  if (title.length === 0 || /[\u0000-\u001f\u007f-\u009f\u2028\u2029\ud800-\udfff]/u.test(title)) {
    return failure('INVALID_TITLE');
  }
  if ([...title].length > MAX_TITLE_LENGTH) return failure('TITLE_TOO_LONG');
  return { ok: true, value: title };
}

export function createTask(input: Readonly<{
  id: unknown;
  title: unknown;
  createdAtEpochMs: unknown;
}>): Result<Task> {
  const title = validateTitle(input.title);
  if (!title.ok) return title;
  if (typeof input.id !== 'string' || !/^[A-Za-z0-9_-]{1,100}$/u.test(input.id)) {
    return failure('INVALID_ID');
  }
  const timestamp = input.createdAtEpochMs;
  if (typeof timestamp !== 'number' || !Number.isSafeInteger(timestamp)
    || timestamp < 0 || timestamp > 8_640_000_000_000_000) {
    return failure('INVALID_TIMESTAMP');
  }
  return {
    ok: true,
    value: Object.freeze({ id: input.id, title: title.value, completed: false, createdAtEpochMs: timestamp }),
  };
}

export function toggleTask(task: Task): Task {
  return Object.freeze({ ...task, completed: !task.completed });
}
